File: src/main/java/project/rpg/Rpg.java
package project.rpg;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import project.rpg.commands.*;
import project.rpg.listeners.*;
import project.rpg.manager.FileManager;
import project.rpg.manager.ItemManager;
import project.rpg.player.PlayerInformation;
import project.rpg.player.info.Mana;
import project.rpg.skill.SkillDic;
import project.rpg.ui.ActionBarUI;

import java.util.Objects;

public final class Rpg extends JavaPlugin {
    public ActionBarUI _actionBar = new ActionBarUI(this);
    public Mana _mana = new Mana(this);
    @Override
    public void onEnable() {

        getLogger().info("RPG plugin loading!");

        registerEvents();
        getCommands();
        loadObjects();

        ItemManager.makeItems();

        checkOnlinePlayers();
        _mana.startManaRefilling();
        _actionBar.startActionBar();

        SkillDic.addAll();

        getLogger().info("RPG plugin loaded!");
    }

    @Override
    public void onDisable() {

        saveObjects();
        getLogger().info("RPG plugin disabled");
    }

    private void getCommands(){
        Objects.requireNonNull(getCommand("quests")).setExecutor(new QuestToggleCommand());
        Objects.requireNonNull(getCommand("savef")).setExecutor(new FileSaveTestCommand());
        Objects.requireNonNull(getCommand("status")).setExecutor(new StatusCommand());
        Objects.requireNonNull(getCommand("test")).setExecutor(new TestCommand());
        Objects.requireNonNull(getCommand("titleTest")).setExecutor(new TitleTestCommand());
        Objects.requireNonNull(getCommand("testItem")).setExecutor(new TestItemCommand());
        Objects.requireNonNull(getCommand("skill")).setExecutor(new SkillCommand());
    }

    private void registerEvents(){
        getServer().getPluginManager().registerEvents(new BlockClickEventListener(), this);
        getServer().getPluginManager().registerEvents(new EntityTakeDamageEventListener(this), this);
        getServer().getPluginManager().registerEvents(new InventoryEventListener(), this);
        getServer().getPluginManager().registerEvents(new PlayerItemUseEventListener(this), this);
        getServer().getPluginManager().registerEvents(new PlayerJoinEventListener(), this);
        getServer().getPluginManager().registerEvents(new PlayerLevelUpListener(), this);
        getServer().getPluginManager().registerEvents(new PlayerQuitEventListener(), this);
        getServer().getPluginManager().registerEvents(new PlayerRespawnEventListener(this), this);
    }

    private void loadObjects() {
        FileManager.makeFile();
        FileManager.getFile();
    }

    private void saveObjects(){
        FileManager.makeFile();
        FileManager.saveFile();
    }

    private void checkOnlinePlayers(){
        Bukkit.getScheduler().scheduleSyncDelayedTask(this, ()->{
            for(Player player : Bukkit.getOnlinePlayers()){
                PlayerInformation.makeInfo(player);
            }
        }, 0);
    }
}

File: src/main/kotlin/project/rpg/Rpg.kt
package project.rpg

import io.github.monun.heartbeat.coroutines.HeartbeatScope
import io.github.monun.kommand.kommand
import kotlinx.coroutines.launch
import org.bukkit.Bukkit
import org.bukkit.NamespacedKey
import org.bukkit.plugin.java.JavaPlugin
import project.rpg.commands.*
import project.rpg.commands.debug.*
import project.rpg.database.Database
import project.rpg.events.listeners.*
import project.rpg.events.listeners.debug.CustomMobAttackEventListener
import project.rpg.player.PlayerInformation
import project.rpg.player.mana.Mana
import project.rpg.ui.text.ActionBarUI


class Rpg : JavaPlugin() {

    init{
        actionBar = ActionBarUI()
        plugin = this
    }

    override fun onLoad() {
        Initializer.initializeAll()

        Database.run {
            connect()
            createTables()
        }
    }

    override fun onEnable() {
        logger.info("RPG plugin loading!")
        checkOnlinePlayers()

        registerEvents()
        registerCommands()
        registerTabCompleter()

        //actionBar.startActionBar()
        actionBar.startCoroutineActionBar()
        Mana.startManaRefilling()

    }

    override fun onDisable() {
        logger.info("saving files...")

        Bukkit.getOnlinePlayers().forEach {
            Database.writeUser(it)
        }

        logger.info("RPG plugin disabled")
    }

    private fun registerCommands() {
        getCommand("quests")!!.setExecutor(QuestToggleCommand())
        //getCommand("saveFile")!!.setExecutor(FileSaveCommand())
        getCommand("status")!!.setExecutor(StatusCommand())
        getCommand("titleTest")!!.setExecutor(TitleTestCommand())
        getCommand("getItem")!!.setExecutor(ItemCommand())
        getCommand("menu")!!.setExecutor(MainMenuCommand())
        getCommand("level")!!.setExecutor(LevelCommand())
        getCommand("craft")!!.setExecutor(CraftCommand())
        getCommand("friend")!!.setExecutor(FriendCommand())
        //getCommand("mob")!!.setExecutor(MobTest())
        getCommand("insertsql")!!.setExecutor(SQLInsertCommand())
        getCommand("serialize")!!.setExecutor(SerializeCommand())

        kommand {

            //debug/test commands
            registerTestCommand()
            registerSampleKommand()
            registerMobCommand()

        }
    }

    private fun registerTabCompleter() {
        getCommand("quests")!!.tabCompleter = QuestToggleCommand()
        getCommand("status")!!.tabCompleter = StatusCommand()
        getCommand("friend")!!.tabCompleter = FriendCommand()
    }

    private fun registerEvents() {
        with(server.pluginManager) {
            registerEvents(BlockClickEventListener(), plugin)
            registerEvents(EntityTakeDamageEventListener(), plugin)
            registerEvents(InventoryEventListener(), plugin)
            registerEvents(PlayerItemUseEventListener(), plugin)
            registerEvents(PlayerJoinEventListener(), plugin)
            registerEvents(PlayerQuitEventListener(), plugin)
            registerEvents(RespawnEventListener(), plugin)
            registerEvents(PlayerItemConsumeEventListener(plugin as Rpg), plugin)
            registerEvents(ProjectileEventListener(), plugin)

            //debugging/testing events
            registerEvents(CustomMobAttackEventListener(), plugin)
        }
    }

    private fun checkOnlinePlayers() {
        HeartbeatScope().launch {
            Bukkit.getOnlinePlayers().forEach {
                PlayerInformation.makeInfo(it)
            }
            logger.info("RPG plugin loading completed")
        }
    }

    companion object{
        lateinit var actionBar: ActionBarUI
        lateinit var plugin: JavaPlugin

        fun pluginNamespacedKey(key: String):NamespacedKey = NamespacedKey(plugin, key)
    }
}